File: Crio/crio_programming_interview_problems-master/ContinuousMedian/ContinuousMedian.py

def continuousMedian(arr):
    # CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
    from heapq import heappush, heappop, heapify
    medians = list()
    maxheap = list()   # max heap
    minheap = list()     # min heap
    for element in arr:
        heappush(maxheap, -element)
        heappush(minheap, -maxheap[0])
        heappop(maxheap)
        if len(maxheap) < len(minheap):
            heappush(maxheap, -minheap[0])
            heappop(minheap)
        if len(maxheap) == len(minheap):
            medians.append( int((-maxheap[0] + minheap[0])/2))
        else:
            medians.append(-maxheap[0])
    return medians
    # CRIO_SOLUTION_END_MODULE_L1_PROBLEMS


def main():
    n= int(input())
    arr = list(map(int,input().split()))
    answer = continuousMedian(arr)
    print('\n'.join(map(str, answer)))

if __name__ == "__main__":
    main()

File: Crio/crio_programming_interview_problems-master/ContinuousMedian/ContinuousMedian.java
import java.util.*;
import java.io.*;

class ContinuousMedian {
	// CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
	// class Cmp implements Comparator<Integer> {
	// 	public int compare(Integer a , Interger b) {
	// 		return a - b;
	// 	}
	// }
	// CRIO_SOLUTION_END_MODULE_L1_PROBLEMS
	
	public static ArrayList<Integer> continuousMedian(ArrayList<Integer> nums) {
		// CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
		int n = nums.size();
		PriorityQueue <Integer> maxHeap = new PriorityQueue<Integer> ();
		PriorityQueue<Integer> minHeap = new PriorityQueue<Integer> (Collections.reverseOrder());
		minHeap.add(nums.get(0));
		double median = nums.get(0);

		ArrayList<Integer> answer = new ArrayList<Integer>();
		answer.add((int)median);


		for ( int i = 1; i < n; i++) {
			int cur = nums.get(i);
			if ( minHeap.size() > maxHeap.size()) {
				if (cur < median) {
					maxHeap.add(minHeap.remove());
					minHeap.add(cur);
				} else {
					maxHeap.add(cur);
				}
				median = (double)(minHeap.peek() + maxHeap.peek()) / 2;
			} else if ( minHeap.size() == maxHeap.size()) {
				if ( cur < median ) {
					minHeap.add(cur);
					median = minHeap.peek();
				} else {
					maxHeap.add(cur);
					median = maxHeap.peek();
				}
			} else {
				if ( cur > median) {
					minHeap.add(maxHeap.remove());
					maxHeap.add(cur);
				} else {
					minHeap.add(cur);
				}
				median = (double)(minHeap.peek() + maxHeap.peek()) / 2;
			}

			answer.add((int)median);
		}
		return answer;
                // CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

	}

	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		int n = sc.nextInt();
		ArrayList<Integer> A = new ArrayList<Integer>();
		for ( int i = 0; i < n; i++) {
			int x = sc.nextInt();
			A.add(x);
		}

		ArrayList<Integer> answer = continuousMedian(A);
		for ( int i = 0; i < answer.size(); i++) {
			System.out.println(answer.get(i) + " ");
		}

	}
}
